pythonwhat/State.py

```diff
@@ -395,5 +203,17 @@ def parse_int(x):
     def set_active_state(state):
         State.active_state = state
 
+# add property methods for retrieving parser outputs
+# e.g. state.student_withs
+from functools import partial
+for k, Parser in parser_dict.items():
+    def getx(self, tree_name, Parser=Parser): 
+        p = Parser()
+        p.visit(getattr(self, tree_name))
+        return p.out
+
+    for s in ['student', 'solution']:
+        setattr(State, s+'_'+k, property(partial(getx, tree_name = s+'_tree')))
```

Comment:
Recap from pairing: the getters are run each time the property is accessed. This means that with the current setup that multiple test_function calls *do* run the parser multiple times. The getters can be modified to cache the results on their state instance (similar to the original code). I'll add a _cache property to the instance that that getters check first.